wgengine/monitor/monitor.go

```diff
@@ -264,3 +303,33 @@ func jsonSummary(x interface{}) interface{} {
 	}
 	return j
 }
+
+func wallTime() time.Time {
+	// From time package's docs: "The canonical way to strip a
+	// monotonic clock reading is to use t = t.Round(0)."
+	return time.Now().Round(0)
+}
+
+func (m *Mon) pollWallTime() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if m.closed {
+		return
+	}
+	m.checkWallTimeAdvanceLocked()
+	if m.timeJumped {
+		m.InjectEvent()
+	}
+	m.wallTimer.Reset(pollWallTimeInterval)
+}
+
+// checkWallTimeAdvanceLocked updates m.timeJumped, if wall time jumped
+// more than 150% of pollWallTimeInterval, indicating we probably just
+// came out of sleep.
+func (m *Mon) checkWallTimeAdvanceLocked() {
+	now := wallTime()
+	if now.Sub(m.lastWall) > pollWallTimeInterval*3/2 {
+		m.timeJumped = true
+	}
+	m.lastWall = now
```

Comment:
I'd think you only want to set m.lastWall when you're re-arming the timer. Otherwise you risk constantly updating lastWall due to other events and missing genuine wallTime overages. (Maybe it doesn't matter in practice, I guess.)

Reply:
I spent a bit of time going back and forth on that but I couldn't convince myself that it mattered. In the end I thought it was simpler to have fewer branches and just always make sure lastWall was any lastWall we'd need from any path, including real network change events, or the upcoming just-poll-time-in-wake method.